### DISCUSSION TITLE
QUIC support

### DISCUSSION THREAD

**Discussion by NKShukla** [mitmproxy/mitmproxy#5807] [Q&A] [answered]
Based on previous discussions [#5499](https://github.com/mitmproxy/mitmproxy/discussions/5499) and [#5535](https://github.com/mitmproxy/mitmproxy/discussions/5535), is there any improvement being made for providing support for the h3/quic requests? I have checked out some recent PRs [#5434](https://github.com/mitmproxy/mitmproxy/pull/5435), and it seems like MITM is now supporting QUIC as well. However, I could not find any documentation regarding how to use it. Also, when I tried running YouTube on Chrome with _mitmweb_ and collected pcaps simultaneously, I did not see any QUIC packet. Can anyone please help as to how I can enable QUIC support?

**Comment by mhils** [comment, marked as answer]
We do  support QUIC and HTTP/3 on the main branch now, but getting QUIC requests through mitmproxy is still a bit tricky. There are two main problems here:

 1. Regular proxying via HTTP CONNECT only supports TCP, but QUIC is a UDP-based protocol. This means that mitmproxy in regular mode won't be able to do QUIC. QUIC only works with transparent, WireGuard, and (to some extent) reverse proxy mode.
 2. It currently looks like some clients, for example Chrome on Android, do not respect user-added CAs when using QUIC. This looks like a bug in Chrome, but we haven't dug into it yet.

We have two things planned to improve the overall situation:

1. Docs. As you said, most stuff is currently still undocumented. `mitmdump --mode reverse:http3://cloudflare-quic.com:443` and `mitmdump --mode reverse:quic://cloudflare-quic.com:443` work, but that's not mentioned anywhere. We're in a very early testing phase really.
2. OS Proxying. We plan to supplement WireGuard mode with an OS-level proxy mode to support single-device settings
    (see https://github.com/mitmproxy/mitmproxy/discussions/5795#discussioncomment-4341372).

**Reply by yashikada** [reply]
> QUIC only works with transparent

are you sure? From the [docs](https://docs.mitmproxy.org/stable/concepts/modes/#transparent-proxy)

Consider using [WireGuard](https://docs.mitmproxy.org/stable/concepts/modes/#wireguard) and [local capture](https://docs.mitmproxy.org/stable/concepts/modes/#local-capture) mode instead of transparent mode. They are easier to set up and also support UDP-based protocols (which transparent mode currently does not).
